modules/apps/frontend-css/frontend-css-variables-web/src/main/java/com/liferay/frontend/css/variables/web/internal/theme/ThemeCSSVariableDescriptionsServiceTrackerCustomizer.java

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.css.variables.web.internal.theme;
+
+import com.liferay.frontend.css.variables.CSSVariableDescription;
+import com.liferay.portal.kernel.json.JSONException;
+import com.liferay.portal.kernel.json.JSONFactory;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.servlet.ServletContext;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+public class ThemeCSSVariableDescriptionsServiceTrackerCustomizer
+	implements ServiceTrackerCustomizer
+		<ServletContext, AtomicReference<Map<String, CSSVariableDescription>>> {
+
+	public ThemeCSSVariableDescriptionsServiceTrackerCustomizer(
+		JSONFactory jsonFactory) {
+
+		_jsonFactory = jsonFactory;
+	}
+
+	@Override
+	public AtomicReference<Map<String, CSSVariableDescription>> addingService(
+		ServiceReference<ServletContext> serviceReference) {
+
+		AtomicReference<Map<String, CSSVariableDescription>> atomicReference =
+			new AtomicReference<>();
+
+		modifiedService(serviceReference, atomicReference);
+
+		return atomicReference;
+	}
+
+	@Override
+	public void modifiedService(
+		ServiceReference<ServletContext> serviceReference,
+		AtomicReference<Map<String, CSSVariableDescription>> atomicReference) {
+
+		Bundle bundle = serviceReference.getBundle();
+
+		if (!ThemeBundleInspector.isTheme(bundle)) {
```

Comment:
Feels like a lot of duplicated `.isTheme` here and in the `ThemeBundleInspector` constructor.

I know it's safer, but the current code path can't throw the exception because you're checking here earlier in a static method to see if the bundle is a theme... so what's the point either in checking here or in throwing there?

Reply:
Defensive programming and failing early. Otherwise, we can create a `ThemeBundleInspector` for a non-theme bundle and call the other methods and they may fail or return bogus values randomly.

I also thought about the duplication but couldn't find a better pattern :thinking:. I tried:

1. The static method to check before construction so that you don't need to rely on a exception to do flow control 
2. Only check in constructor: you don't check twice, but the only way to test if a bundle is a theme is to construct and catch exceptions
3. An `isTheme` method in the object. This is cleaner, but then you need to check `isTheme` in every other method to avoid returning invalid values.

Another possibility could be a factory method to check and construct the object so that if the bundle is not a theme it returns null.

I don't know which one I prefer. Also I'm not sure which one will pass further revisions better...